osu.Server.Spectator/Hubs/ScoreUploader.cs

```diff
@@ -0,0 +1,163 @@
+            cancellation.CancelAfter(TimeSpan.FromMilliseconds(TimeoutInterval));
+
+            queue.Enqueue(new UploadItem(token, score, cancellation));
+        }
+
+        /// <summary>
+        /// Flushes all pending uploads.
+        /// </summary>
+        public async Task Flush()
+        {
+            try
+            {
+                timer.Stop();
+
+                if (queue.IsEmpty)
+                    return;
+
+                Console.WriteLine($"Beginning upload of {queue.Count} scores");
+
+                using (var db = databaseFactory.GetInstance())
+                {
+                    int countToTry = queue.Count;
+
+                    for (int i = 0; i < countToTry; i++)
+                    {
+                        if (!queue.TryDequeue(out var item))
+                            continue;
+
+                        long? scoreId = await db.GetScoreIdFromToken(item.Token);
+
+                        if (scoreId == null && !item.Cancellation.IsCancellationRequested)
+                        {
+                            // Score is not ready yet - enqueue for the next attempt.
+                            queue.Enqueue(item);
+                            continue;
+                        }
+
+                        try
+                        {
+                            if (scoreId != null)
+                            {
+                                item.Score.ScoreInfo.OnlineID = scoreId.Value;
+                                await scoreStorage.WriteAsync(item.Score);
+                            }
+                            else
+                                Console.WriteLine($"Score upload timed out for token: {item.Token}");
+                        }
+                        finally
+                        {
+                            item.Dispose();
+                            Interlocked.Decrement(ref remainingUsages);
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"Error during score upload: {e}");
+            }
+            finally
+            {
+                if (timerCancellationToken.IsCancellationRequested)
+                    timer.Dispose();
+                else
+                    timer.Start();
+            }
+        }
+
+        public void Dispose()
+        {
+            timerCancellationSource.Cancel();
+            timerCancellationSource.Dispose();
+        }
+
+        private record UploadItem(long Token, Score Score, CancellationTokenSource Cancellation) : IDisposable
+        {
+            public void Dispose()
+            {
+                Cancellation.Dispose();
+            }
+        }
+
+        private int remainingUsages;
+
+        // Using the count of items in the queue isn't correct since items are dequeued for processing.
+        public int RemainingUsages => remainingUsages;
```

Comment:
Probably use `TotalInFlight` to match `osu-queue-processor` terminology.

Reply:
Hmmm, this is just because it implements `IEntityStore`. Perhaps I should use a different interface?

Reply:
Let's leave it for now.